druid/examples/calc.rs

```diff
@@ -112,95 +115,239 @@ impl CalcState {
     }
 }
 
-fn pad<T: Data>(inner: impl Widget<T> + 'static) -> impl Widget<T> {
+#[derive(Debug)]
+struct ButtonInfo {
+    row_id: usize,
+    col_id: usize,
+}
+
+#[derive(PartialEq, Eq, Hash, Debug)]
+enum ButtonId {
+    Op(char),
+    Digit(u8),
+}
+
+struct KeyMap {
+    // Note we rely on the fact that the key_map has row and column ids
+    // into the Flex, so the flex here cannot be dynamically changing.
+    key_map: HashMap<ButtonId, ButtonInfo>,
+    flex: Flex<CalcState>,
+}
+
+impl KeyMap {
+    fn new(flex: Flex<CalcState>, key_map: HashMap<ButtonId, ButtonInfo>) -> KeyMap {
+        KeyMap { key_map, flex }
+    }
+}
+
+fn key_event_to_button_id(event: &KeyEvent) -> Option<ButtonId> {
+    use crate::ButtonId::*;
+    match event.mods {
```

Comment:
for this sort of thing, look at the `HotKey` type; basically you can describe a key + mods, and then see if an event matches it. For an example, see how we handle keys in [the textbox widget](https://github.com/xi-editor/druid/blob/master/druid/src/widget/textbox.rs#L380)